File: export.js
#!/usr/bin/env node

var z = require('zombie')
  , fs = require('fs')
  , _ = require('underscore')

function error(e) { console.log(e) }

function getCredentials(next) {

    fs.readFile('.credentials', function(err, data) {

        if (err) {
            next(err)
            return
        }

        var pair = data.toString('ascii').split('\n')

        if (pair.length < 2) {
            next('Wrong credentials file.')
            return
        }

        next(null, {username: pair[0], password: pair[1]})
    })
}

function log(br) {
    console.log(br.location._url.href)
    console.log(br.statusCode)
    //console.log(br.html())
}

function elemExists(br, query) {
    return br.query(query) !== undefined
}

function openLoginPage(next) {

    var br = new z()//{debug: true})
      , stages = [
            {
                name: 'start',
                action: function(br) {
                    br.visit('https://28degrees-online.gemoney.com.au/')
                }
            }
          , {
                name: 'intermediate',
                urlPart: 'https://28degrees-online.gemoney.com.au/access/do?',
                check: function(br) {
                    return true
                },
                action: function(br) {
                    if (br.html().indexOf("window.location = '/access/login'") !== -1) {
                        br.location = '/access/login'
                    }
                }
            }
          , {
                name: 'login',
                urlPart: 'https://28degrees-online.gemoney.com.au/access/login',
                count: 0,
                check: function(br) {
                    /*var error = elemExists(br, ':contains("There seems to be a problem")')
                    if (error) {
                        exit = true
                        console.log(br.html())
                        return false
                    }

                    if (_exit) {
                        console.log(br.html())
                        exit = true
                    }*/

                    this.count += 1
                    if (this.count > 1) { // This page have been displayed more than once, login error
                        console.log('Error logging in. Check your username/password in .credentials file')
                        return false
                    }

                    return br.query('input[name="SUBMIT"]') !== undefined
                },
                action: function(br) {
                    console.log('Logging in...')
                    br.document.getElementById('AccessToken_Username').value = creds.username
                    br.document.getElementById('AccessToken_Password').value = creds.password
                    br.pressButton('[name="SUBMIT"]')
                }
            }
          , {
                name: 'home',
                urlPart: 'https://28degrees-online.gemoney.com.au/wps/myportal/ge28degrees/public/home',
                check: function(br) {
                    return elemExists(br, ':contains("My Recent Transactions")')
                },
                action: function(br) {
                    var h = br.query('h2:contains("My Recent Transactions")')

                    if (!h) {
                        console.log('Cannot find "My Recent Transactions" section on home page. Requesting exit.')
                        this.exit = true
                        return
                    }

                    var link = h.parentNode.parentNode.querySelector('a[name="Wrapper_lnMoreInfo"]')
                    if (!link) {
                        console.log('Cannot find link "MORE..." in "My Recent Transactions" section. Requesting exit.')
                        this.exit = true
                        return
                    }

                    br.fire('click', link)
                }
            }
          , {
                name: 'transactions',
                urlPart: 'https://28degrees-online.gemoney.com.au/wps/myportal/ge28degrees/public/account/transactions/',
                check: function(br) {
                    return elemExists(br, 'tr[name="DataContainer"]')
                },
                action: function(br) {

                    var rows = br.queryAll('tr[name="DataContainer"]')
                      , isNextButtonVisible = elemExists(br, 'a[name="nextButton"]')

                    console.log('Rows: ' + rows.length)
                    console.log('Dates: ' + br.queryAll('[name="Transaction_TransactionDate"]').length)
                    console.log('Dates 2: ' + br.queryAll('td', rows[0]).length)

                    _(rows).each(function(row) {

                        var cells = row.querySelectorAll('td')
                          , date
                          , name
                          , desc
                          , amt

                        _(cells).each(function(cell) {


                            if (cell.childNodes.length !== 0 &&
                                cell.childNodes[0].tagName.toLowerCase() === 'span') {

                                var span = cell.childNodes[0]

                                //console.log('span:' + span.attributes.getNamedItem('name').value)
                                //console.log('span:' + span.innerHTML)

                                switch (span.attributes.getNamedItem('name').value) {
                                    case 'Transaction_TransactionDate': date = span.innerHTML;
                                    case 'Transaction_CardName': name = span.innerHTML;
                                    case 'Transaction_TransactionDescription': desc = span.innerHTML;
                                    case 'Transaction_Amount': amt = span.innerHTML;
                                }
                            }
                        })

                        //console.log(row.innerHTML)
                        //console.log(date.innerHTML + ':' + name.innerHTML + ':' + desc.innerHTML + ':' + amt.innerHTML)

                    })

                    if (isNextButtonVisible) {
                        console.log('Some records available, going further back...')
                        var link = br.query('a[name="nextButton"]')
                        console.log('Link: ' + link.attributes.getNamedItem('href').value)
                        br.fire('click', link)
                    } else {
                        console.log('Looks like we have reached the end of transactions. Requesting exit.')
                        this.exit = true
                    }
                }
            }
        ]


    br.on('loaded', function(br) {

        var url = br.location._url.href

        console.log(url)
        stage = _(stages).find(function(stage) {
            return url.indexOf(stage.urlPart) === 0
        })

        if (!stage || !stage.check.call(stage, br)) {
            return
        } else {
            stage.action.call(stage, br)
            if (typeof stage.exit !== 'undefined' && stage.exit) {
                console.log('Stage "' + stage.name + '" requested exit, exiting...')
                return
            }
        }
    })

    stages[0].action(br)

    /*
    br.clickLink('My Account', function(err, br, status) {

        if (err) {
            next(err)
            return
        }

        br.clickLink("Transactions", function(err, br, status) {

            if (err) {
                next(err)
                return
            }

            //log(br)
            console.log(br.html())

        })
    })*/

    br.on('done', function() {
        console.log('Done!')
        console.log(arguments)
    })

    /*
    br.on('error', function() {
        console.log('Error!')
        console.log(arguments)
    })
    */
/*
    br.visit('https://28degrees-online.gemoney.com.au/', function(e, br) {

        if (e) {
            next(e)
            return
        }

        br.visit('https://28degrees-online.gemoney.com.au/access/login', function(e, br) {

            if (e) {
                next(e)
                return
            }

            br.document.getElementById('AccessToken_Username').value = creds.username
            br.document.getElementById('AccessToken_Password').value = creds.password

            console.log('PRESSING')
            br.pressButton('[name="SUBMIT"]', function(err, br) {

                if (err) {
                    next(err)
                    return
                }

            })

            //br.wait(isLoaded, function() {

            //})



        })


    })*/
}

var creds = null

getCredentials(function(err, data) {

    if (err) {
        error(err)
        return
    }

    creds = data
    openLoginPage(function(err) {
        console.log(err)
    })
})

File: export.py
#!/usr/bin/env python

from __future__ import print_function
from getpass import getpass
import re
from datetime import datetime

from mechanize import Browser
from pyquery import PyQuery
from collections import namedtuple

import db


Transaction = namedtuple('Transaction',
                         ['date', 'payer', 'amount', 'memo', 'payee'])


def messages(before, after_ok, after_fail):

    def external_decorator(f):
        def wrapped(*args, **kwargs):
            print(before)
            r = f(*args, **kwargs)
            print(after_ok if r else after_fail)
            return r
        return wrapped

    return external_decorator


def get_credentials():

    print('Enter your username and password: ')
    lines = []
    lines.append(raw_input())
    lines.append(getpass())

    return lines


def get_node_text(node):
    return node.text if len(node.text) != 0 else None


"""
0-22 payee, 23-37 loc, 38-$ loc
PAYPAL *KOBO INC       XXXXXXXXXX    ON
WWW.THREADLESS.COM     XXXXXXXXXXX   IL
"""
def fetchTransactions(text):

    q = PyQuery(text)
    trans = []

    for row in q('tr[name="DataContainer"]'):

        date = get_node_text(q('span[name="Transaction_TransactionDate"]', row)[0])
        payer = get_node_text(q('span[name="Transaction_CardName"]', row)[0])
        desc_payee = get_node_text(q('span[name="Transaction_TransactionDescription"]', row)[0])
        amount = get_node_text(q('span[name="Transaction_Amount"]', row)[0])

        if len(desc_payee) >= 23:
            payee = desc_payee[:23]
            memo = desc_payee[23:]
        else:
            payee = desc_payee
            memo = ''

        # Clean up the data
        amount = amount.replace('$', '')
        payee = re.sub('\s+', ' ', payee)
        memo = re.sub('\s+', ' ', memo)

        trans.append(Transaction(date=date,
                                 payer=payer,
                                 amount=amount,
                                 memo=memo,
                                 payee=payee))

    return trans


"""See http://en.wikipedia.org/wiki/Quicken_Interchange_Format for more info."""
@messages('Writing QIF file...', 'OK', '')
def write_qif(trans):

    f_str = '%d/%m/%Y'
    s_d = datetime.strptime(reduce(lambda t1, t2: t1 if datetime.strptime(t1.date, f_str) <
                                                  datetime.strptime(t2.date, f_str) else t2,
                                   trans).date, f_str)
    e_d = datetime.strptime(reduce(lambda t1, t2: t1 if datetime.strptime(t1.date, f_str) >
                                                  datetime.strptime(t2.date, f_str) else t2,
                                   trans).date, f_str)

    out_str = '%Y.%m.%d'
    file_name = './export/%s-%s.qif' % (s_d.strftime(out_str), e_d.strftime(out_str))
    with open(file_name, 'w') as f:

        # Write header
        print('!Account', file=f)
        print('NQIF Account', file=f)
        print('TCCard', file=f)
        print('^', file=f)
        print('!Type:CCard', file=f)

        for t in trans:
            print('C', file=f) # status - uncleared
            print('D' + t.date, file=f) # date
            print('T' + t.amount, file=f) # amount
            print('M' + t.payer + ' ' + t.memo, file=f) # memo
            print('P' + t.payee, file=f) # payee
            print('^', file=f) # end of record


@messages('Writing CSV file...', 'OK', '')
def write_csv(trans):

    f_str = '%d/%m/%Y'
    s_d = datetime.strptime(reduce(lambda t1, t2: t1 if datetime.strptime(t1.date, f_str) <
                                                  datetime.strptime(t2.date, f_str) else t2,
                                   trans).date, f_str)
    e_d = datetime.strptime(reduce(lambda t1, t2: t1 if datetime.strptime(t1.date, f_str) >
                                                  datetime.strptime(t2.date, f_str) else t2,
                                   trans).date, f_str)

    out_str = '%Y.%m.%d'
    file_name = './export/%s-%s.csv' % (s_d.strftime(out_str), e_d.strftime(out_str))
    with open(file_name, 'w') as f:
        print('Date,Amount,Payer,Payee', file=f)
        for t in trans:
            print('"%s","%s","%s","%s"' % (t.date, t.amount, t.payer, t.payee), file=f)


@messages('Logging in...', 'OK', 'Login failed')
def login(creds):

    br = Browser()

    br.open('https://28degrees-online.gemoney.com.au/')
    br.open('https://28degrees-online.gemoney.com.au/access/login')

    br.select_form(nr=0)
    br.form['USER'] = creds[0]
    br.form['PASSWORD'] = creds[1]
    br.submit()

    text = br.response().read()
    if "window.location = '/access/login';" in text:
        return None

    return br


@messages('Opening transactions page...', 'OK', 'Exiting...')
def open_transactions_page(br):

    br.open('https://28degrees-online.gemoney.com.au/wps/myportal/ge28degrees/public/account/transactions/')
    text = br.response().read()

    if 'New card number required' in text:
        q = PyQuery(text)
        cancel_btn = q('input[name="cancelButton"]')

        if len(cancel_btn) == 0:
            print('No cancel button found on "New card required" page')
            return None

        cancel_btn = cancel_btn[0]
        matches = re.match('location\.href="(.*)"', cancel_btn.attrib['onclick'])

        if len(matches.groups()) == 0:
            print('No onclick event in cancel button found')
            return None

        # Cancel new card number submission
        br.open('https://28degrees-online.gemoney.com.au' + matches.groups()[0])
        br.open('https://28degrees-online.gemoney.com.au/wps/myportal/ge28degrees/public/account/transactions/')

    return br


def export():

    t_db = db.init_db()
    if not t_db:
        print('Error initialising database')
        return

    creds = get_credentials()
    if not creds:
        return

    br = login(creds)
    if not br:
        return

    br = open_transactions_page(br)
    if not br:
        return

    trans = []

    while True:
        text = br.response().read()

        q = PyQuery(text)

        page_trans = fetchTransactions(text)
        trans += page_trans

        nextButton = q('a[name="nextButton"]')
        isNextVisible = len(nextButton) != 0
        if not isNextVisible:
            break

        page_count = len(page_trans)
        print('Got %s transactions, from %s to %s' % (page_count,
                                                      page_trans[0].date,
                                                      page_trans[-1].date))
        print('Opening next page...')
        br.open(nextButton[0].attrib['href'])

        #if len(trans) > 50:
        #    break

    new_trans = db.get_only_new_transactions(trans)
    print('Total of %s new transactions obtained' % len(new_trans))

    if len(new_trans) != 0:
        print('Saving transactions...')
        db.save_transactions(new_trans)

        write_qif(new_trans)
        write_csv(new_trans)


if __name__ == "__main__":
    export()
